controllers/utils.js

```diff
@@ -33,71 +33,161 @@ function hasUserSignedUpForRelay(user) {
   return false;
 }
 
-function getExperimentBranch(req, sorterNum = false, language = false) {
+function chooseVariation(variations, sorterNum){
 
-  if (sorterNum === false) {
-    sorterNum = Math.floor(Math.random() * 100);
-    log.debug("No coinflip number provided. Coinflip number is ", sorterNum);
+  let totalPercentage = 0;
+
+  // calculate and store total percentage of variations
+  for (const v in variations) {
+    if (variations.hasOwnProperty(v) && typeof variations[v] === "number") {
+      // multiply by 100 to allow for percentages to the hundredth
+      // (and avoid floating point math errors)
+      totalPercentage += (variations[v] * 100);
+    }
+  }
+
+  totalPercentage = totalPercentage / 100;
+
+  // Make sure totalPercent is between 0 and 100
+  if (totalPercentage === 0 || totalPercentage > 100) {
+    throw new Error(`The total percentage ${totalPercentage} is out of bounds!`);
   }
 
-  if (req.session.excludeFromExperiment && !req.query.experimentBranch) {
+  // make sure random number falls in the distribution range
+  let runningTotal;
+  let choice;
+
+  if (sorterNum <= totalPercentage) {
+      runningTotal = 0;
+
+      // loop through all variations
+      for (const v in variations) {
+          // check if random number falls within current variation range
+          if (sorterNum <= (variations[v] + runningTotal)) {
+              // if so, we have found our variation
+              choice = v;
+              break;
+          }
+
+          // tally variation percentages for the next loop iteration
+          runningTotal += variations[v];
+      }
+    }
+
+    return choice;
+
+}
+
+function getExperimentBranch(req, sorterNum = false, language = false, variations) {
```

Comment:
Since this gets the experiment branch AND sets it into the session, can we re-name this to `setExperimentBranch`?